library/Backends/llvm/src/Juvix/Backends/LLVM/Compilation.hs

```diff
@@ -1,38 +1,47 @@
-{-# LANGUAGE NamedFieldPuns #-}
-{-# LANGUAGE OverloadedStrings #-}
-
 module Juvix.Backends.LLVM.Compilation
   ( compileProgram,
   )
 where
 
-import qualified Data.String as S (fromString)
+import qualified Juvix.Backends.LLVM.Codegen.Block as Block
+import qualified Juvix.Backends.LLVM.Codegen.Types as Types
 import Juvix.Backends.LLVM.Primitive
 import qualified Juvix.Core.Erased.Ann as ErasedAnn
 import Juvix.Library
-import Juvix.Library.Feedback
-import qualified Juvix.Library.HashMap as Map
+import qualified Juvix.Library.Feedback as Feedback
 import qualified Juvix.Library.NameSymbol as NameSymbol
-import qualified LLVM.AST as LLVM (Module, Name, Operand (..))
+import qualified LLVM.AST as LLVM (Operand (..))
 import qualified LLVM.AST.Constant as LLVM (Constant (..))
 import qualified LLVM.AST.Type as LLVM
-import qualified LLVM.IRBuilder as LLVM
 import qualified LLVM.Pretty as LLVM
 import qualified Prelude as P
 
--- | A mapping between Juvix variable names and their operands in LLVM.
-type Env = Map.Map NameSymbol.T LLVM.Operand
+--------------------------------------------------------------------------------
+-- Top Level Compilation
+--------------------------------------------------------------------------------
 
 -- | Compile the input program to an LLVM module.
 -- TODO: maybe do something smarter with the module name?
 compileProgram ::
   Monad m =>
   -- | Term to compile.
   ErasedAnn.AnnTerm PrimTy RawPrimVal ->
-  FeedbackT [] P.String m Text
-compileProgram t = do
-  let llvmmod = LLVM.buildModule "juvix-module" $ mkMain t
-  return $ toStrict $ LLVM.ppllvm llvmmod
+  Feedback.FeedbackT [] P.String m Text
+compileProgram t =
+  case Block.runEnvState (mkMain t) mempty of
+    (Right _, state) ->
+      state
+        |> Types.moduleAST
+        |> LLVM.ppllvm
+        |> toStrict
+        |> return
+    (Left err, _) -> do
+      show err
+        |> Feedback.fail
```

Comment:
Is there any way for the codegen to fail, without it being a bug in the codegen?

In my mind feedback is only used for errors where the user is involved. Regular `error` is for compiler implementation errors. For example a failed lookup on a variable should not never happen in the backends, and is therefore an `error` instead of a `Feedback`. What do you think?

Reply:
How else do you express errors to the user? It seems like `feedback` is the only mechanism we can give back here. We don't return an either, and the way it's used is used for the pipleine itself does give fatal errors IIRC.